src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/decode/ValueDecoder.scala

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2016-2019 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.analytics.scalasdk
+package decode
+
+// java
+import java.time.Instant
+import java.time.format.DateTimeParseException
+import java.util.UUID
+
+// cats
+import cats.syntax.either._
+import cats.syntax.option._
+import cats.syntax.show._
+
+// iglu
+import com.snowplowanalytics.iglu.core.SelfDescribingData
+import com.snowplowanalytics.iglu.core.circe.instances._
+
+// circe
+import io.circe.parser.{parse => parseJson}
+import io.circe.{Error, Json}
+
+// This library
+import com.snowplowanalytics.snowplow.analytics.scalasdk.Common.{ContextsCriterion, UnstructEventCriterion}
+import com.snowplowanalytics.snowplow.analytics.scalasdk.SnowplowEvent.{Contexts, UnstructEvent}
+
+private[decode] trait ValueDecoder[A] {
+  def parse(column: (Key, String)): DecodedValue[A]
+}
+
+private[decode] object ValueDecoder {
+  def apply[A](implicit readA: ValueDecoder[A]): ValueDecoder[A] = readA
+
+  def fromFunc[A](f: ((Key, String)) => DecodedValue[A]): ValueDecoder[A] =
+    new ValueDecoder[A] {
+      def parse(field: (Key, String)): DecodedValue[A] = f(field)
+    }
+
+  implicit final val stringColumnDecoder: ValueDecoder[String] =
+    fromFunc[String] {
+      case (key, value) =>
+        if (value.isEmpty) (key, s"Field $key cannot be empty").asLeft else value.asRight
+    }
+
+  implicit final val stringOptionColumnDecoder: ValueDecoder[Option[String]] =
```

Comment:
can't we define `ValueDecoder[Option[A]]` and the necessary `ValueDecoder[A]`?

Reply:
I don't think there's an elegant way to make a generic `Option` decoder covering the ones below (there are some differences in `Boolean` and `Instant` decoding in particular).